woocommerce/google-listings-and-ads
# Update create campaigns API with multiple targeted locations

### Changes proposed in this Pull Request:

Part of #1248, this PR extends the `POST ads/campaigns` API so the client can provide `targeted_locations` in the request payload to target a campaign in multiple countries.

Some changes that are worth mentioning are:

- Previously the `POST ads/campaigns` API accepts a `country` parameter in the request payload and set it in [Campaign.ShoppingSetting](https://developers.google.com/google-ads/api/reference/rpc/v9/Campaign.ShoppingSetting)'s `sales_country` field. The new API remove the `country` parameter in the request payload and set the **_the country of store's address_** as the country of sale. Ref: `pcTzPl-bo-p2#comment-502`.
- When creating campaign criterion for a campaign in `src/API/Google/AdsCampaignCriterion.php`, I originally added `resource_name` in the create operation like what `AdsGroup.php` did with a unique temporary ID, but the Google Ads API would return an error: `"INCONSISTENT_FIELD_VALUES": "The field's contents don't match another field that represents the same data."`. The error disappeared after removing `resource_name` for a campaign criterion and looks like there were no problems for creating multiple campaigns with multiple campaign criteria. I'm unsure why this happened but found that the [campaign criterion API doesn't have `resource_name` field](https://developers.google.com/google-ads/api/fields/v9/campaign_criterion
) while `AdsGroup` or `CampaignBudget` has `resource_name`.

### Screenshots:

#### Create a new campaign

<img width="1357" alt="Screenshot 2022-03-14 at 17 53 46" src="https://user-images.githubusercontent.com/914406/158148917-c3162aad-8413-4170-a325-2f014a5df540.png">

#### The newly created campaign is included in `GET ads/campaigns` API

<img width="1362" alt="Screenshot 2022-03-14 at 17 58 24" src="https://user-images.githubusercontent.com/914406/158149624-8624cb93-c383-4d53-a212-1db69f833a2b.png">

#### The newly created campaign is returned in `GET ads/campaigns/:campaign-id` API

<img width="1359" alt="Screenshot 2022-03-14 at 17 58 58" src="https://user-images.githubusercontent.com/914406/158149680-afaee0b4-a9ed-4558-ba36-ee8e014aeea7.png">

#### The newly created campaign is displayed in Google Ads dashboard

<img width="1789" alt="Screenshot 2022-03-14 at 17 57 28" src="https://user-images.githubusercontent.com/914406/158149740-360bbd9f-eb84-40a1-a2b8-8ae640c31fa1.png">

### Detailed test instructions:


1. Create a campaign by requesting to `POST ads/campaigns` using Postman like the screenshots in the `Screenshots` section. Remember to add your `wp-cookie` and `x-wp-nonce` in the request headers. 
    - To get the value of `wp-cookie` and `x-wp-nonce`, go to `Marketing -> Google Listings & Ads`, open Chrome dev tool's Network tab, clicks any requests to WP, e.g. `wc/gla/ads/campaigns`, you can get those values in the request headers.
2. Request to `GET ads/campaigns` and the newly created campaign should be included in the response data.
3. Get the campaign ID from the above response, and request to `GET ads/campaigns/:campaign-id`, the newly created campaign should be returned.
4. Go to Google Ads dashboard and check if the newly campaign exists with correct targeted locations.

### Changelog entry

> Update - Extend create campaigns API to accept multiple targeted locations
